Call anti-cheating code less often

Dantemss opened the pull request:
- Anti-cheating is now based on both due and feedback dates
- Skip anti-cheating (checking all other assignments) when receiving a task without due/feedback dates
- Anti-cheating only checks tasks that still need PEs or SPEs
- Mark SPEs for recalculation when they have already been used as PEs elsewhere

This should greatly speed up fetch_course_events

Dantemss committed:
Anti-cheating is now based on both due and feedback dates
Skip anti-cheating (checking all other assignments) for tasks without due/feedback dates

Dantemss committed:
Index assigned exercises on exercise_uuid

nathanstitt commented on app/domain/services/upload_student_exercises/service.rb:
```diff
@@ -95,14 +95,15 @@ def process
+                                      .or(Assignment.where(aa[:feedback_at].gt(start_time)))
+                                      .where(student_uuid: student_uuids)
+                                      .pluck(:student_uuid, :assigned_exercise_uuids)
 
```
just double checking, but we do have indexes on `due_at` and `feedback_at` right?

Dantemss replied:
Yeah we have some indexes for them.

Dantemss committed:
Recalculate only assignments that need PEs or SPEs
Activate anti-cheating for assignments that are not yet due, even if feedback is available
Mark SPEs for recalculation when they are taken, instead of deleting them

Dantemss committed:
Post-rebase fixes
